File: Day06/baekjoon-4963-stack.py
stack=[]

# 방향 배열 (상, 우상, 우, 우하, 하, 좌하, 좌, 좌상) - 8방향 
dx = [-1, -1, 0, 1, 1, 1, 0, -1]
dy = [0, 1, 1, 1, 0, -1, -1, -1]

def dfs(start_x, start_y):
    stack=[(start_x, start_y)]
    visited[start_x][start_y]=True

    while stack:
        x, y = stack.pop()

        for i in range(8):
            nx = x+dx[i]
            ny = y+dy[i]

            if 0<= nx < h and 0<=ny<w and not visited[nx][ny] and current_map[nx][ny]==1:
                stack.append((nx, ny))
                visited[nx][ny]=True

while(True):
    w, h = map(int, input().split())
    current_map=[]
    if w==0 and h==0:
        break

    # map 입력 받기
    for i in range(h):
        row = list(map(int, input().split()))[:w]
        current_map.append(row)

    # 방문여부 확인 리스트 준비
    visited = [[False]*w for _ in range(h)]

    island_count =0

    # 모든 지역에 대해 DFS 수행
    for i in range(h):
        for j in range(w):
            if not visited[i][j] and current_map[i][j] == 1:
                dfs(i, j)
                island_count += 1
    

    print(f"{island_count}")

File: Day06/baekjoon-4963-stack.cpp
#include <iostream>
#include <stack>
#include <algorithm>
#include <vector>
using namespace std;

int w, h;
int current_map[50][50];
bool visited[50][50];
int dx[] = {-1, -1, -1, 0, 0, 1, 1, 1};
int dy[] = {-1, 0, 1, -1, 1, -1, 0, 1};

void dfs(int start_x, int start_y)
{
    stack<pair<int, int>> stack;
    stack.push({start_x, start_y});
    visited[start_x][start_y] = true;

    while (!stack.empty())
    {
        int x = stack.top().first;
        int y = stack.top().second;
        stack.pop();

        for (int i = 0; i < 8; i++) // 8방향으로 변경
        {
            int nx = x + dx[i];
            int ny = y + dy[i]; // y로 수정

            if (nx >= 0 && nx < h && ny >= 0 && ny < w &&
                !visited[nx][ny] && current_map[nx][ny] == 1)
            {
                visited[nx][ny] = true; // == 대신 = 사용
                stack.push({nx, ny});
            }
        }
    }
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    while (true)
    {
        cin >> w >> h;
        if (w == 0 && h == 0)
            break;

        // 맵 입력
        for (int i = 0; i < h; i++)
        {
            for (int j = 0; j < w; j++)
            {
                cin >> current_map[i][j];
            }
        }

        // 방문 여부 초기화
        fill(&visited[0][0], &visited[0][0] + 50 * 50, false);

        int island_count = 0;

        // 섬 탐색
        for (int i = 0; i < h; i++)
        {
            for (int j = 0; j < w; j++)
            {
                if (!visited[i][j] && current_map[i][j] == 1)
                {
                    dfs(i, j);
                    island_count++;
                }
            }
        }

        cout << island_count << '\n';
    }

    return 0;
}